stackrox/stackrox
# ROX-23570: Factor out mailing lists from vulnerability reporting

## Description

Prerequisite to reuse form and view for `NotifierConfigurations` (also known as delivery destinations) from Vulnerability Reports in Compliance Reports.

```ts
export type NotifierConfiguration = {
    emailConfig: {
        notifierId: string;
        mailingLists: string[];
        customSubject: string;
        customBody: string;
    };
    notifierName: string;
};
```

Add src/Components/NotifierConfiguration folder following previous refactoring:
* src/Components/EmailNotifier
* src/Components/EmailTemplate

### Changes in EmailTemplate folder

1. Edit EmailTemplateForm.tsx file.
    * Add `isReadOnly` prop to reuse form for view.
    * Add `readOnlyVariant` prop to `TextInput` and `TextArea` elements.
    * Add conditional rendering for Reset links.
2. Edit EmailTemplateModal.tsx file.
    * Use `onChange={null}` to reuse modal for view.

### Changes in NotifierConfiguration folder

1. Create NotifierConfigurationForm.tsx file.
    * Caller provides name of property in formik values via `fieldIdPrefix` prop.
    * Caller updates other form state if needed via optional `onDeleteLastNotifierConfiguration` prop.
        Vulnerability Reports resets the schedule.
    * Component updates formik state via `setFieldValue` prop.
2. Move and rename NotifierConfigurationView.tsx file.
    * Replace `formValues: ReportFormValues` with `notifierConfiguraton: NotifierConfiguration` prop.
    * Replace references to `reportParameters` with optional `renderTemplatePreview` prop for presentation that is specific to vulnerability reports.
    * Replace description list with table to solve alignment problems.
    * Replace generic `Modal` with specific `EmailTemplateModal` element.
3. Move and rename NotifierMailingLists.tsx file.
    * Caller updates form state via `setMailingLists` and `setNotifier` props.
        It was a judgement call for NotifierMailingLists component to delegate its updates to NotifierConfigurationForm componen
    * Separate `updateNotifierList` from `useEffect` hook to solve timing problem (see testing step 1).

### Changes in VulnerabilityReporting folder

Use optional props to separate vulnerability-specific dependencies on form values other than notifier configurations.

1. Edit DeliveryDestinationsForm.tsx file.
    * Factor out reusable code into NotifierConfigurationForm component.
2. Edit ReportReviewForm.tsx file.
3. Delete useEmailTemplateModal.ts file and use conditional rendering instead.
4. Edit ReportJobs.tsx file.
5. Edit ViewVulnReportPage.tsx file.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js 0 = 4633606 - 4633606
        total -3731 = 11574067 - 11577798
    * `ls -al build/static/js/*.js | wc`
        files 0 = 172 - 172
3. `yarn start` in ui

### Manual testing

Prerequisite: create at least one collection.

Create e-mail notifiers on the fly.

1. Create a report and create email notifier
    * Verify payload and response of POST /v1/notifiers request
    * See that notifier is selected automatically. Baseline code intended but did not succeed because of timing problem `filter` method call before response to GET /v1/notifiers request.
    * Verify review details.
    * Verify payload and response of POST /v2/reports/configurations request
2. Edit the first report and ditto items 2 and 3 above.
3. Clone the first report and ditto items 2 and 3 above.

ReportReviewForm

Misaligned in description lists before changes
![Details_alignment](https://github.com/stackrox/stackrox/assets/11862657/c88b15f4-f210-4c4c-ab32-ecc6f33835f4)

Aligned in table after changes
![ReportReviewForm](https://github.com/stackrox/stackrox/assets/11862657/1c5b5880-55b5-41f5-b90e-c2fdab066b71)

ViewVulnReportPage
![ViewVulnReportPage](https://github.com/stackrox/stackrox/assets/11862657/2ff6c0c3-aa81-436d-ae12-9cda4b30d4ea)

### Integration testing

Uh, oh! No tests. Yet.
